File: src/org/genyris/dl/Triple.java
package org.genyris.dl;

import org.genyris.core.*;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.Environment;

public class Triple extends Atom implements Comparable {

    public final Symbol subject;
    public final Symbol predicate;
    public final Exp object;

    public Triple(Symbol subject, Symbol predicate, Exp object) {
        this.subject = subject;
        this.predicate = predicate;
        this.object = object;
    }

    public void acceptVisitor(Visitor guest) throws GenyrisException {
        guest.visitTriple(this);

    }
    
	public Environment makeEnvironment(Environment parent) throws GenyrisException {
		return new TripleEnvironment(parent, this);
	}

    public String toString() {
        return "(triple " + subject + " " + predicate +  " " + object + ")";
    }

    public Symbol getBuiltinClassSymbol(Internable table) {
        return table.TRIPLE();
    }

    public int hashCode() {
        return subject.hashCode() + predicate.hashCode() + object.hashCode();
    }

    public boolean equals(Object compare) {
        if (!(compare instanceof Triple)) {
            return false;
        }
        Triple t = (Triple) compare;
        return subject.equals(t.subject) && predicate.equals(t.predicate) && object.equals(t.object);
    }
	public Exp eval(Environment env) throws GenyrisException {
		return this;
	}

	public int compareTo(Object O) {
        if(!(O instanceof Triple)){
            return -1;
        }
        Triple other = (Triple) O;
        int comparison = subject.compareTo(other.subject);
        if(comparison != 0) {
            return comparison;
        }
        comparison = predicate.compareTo(other.predicate);
        if(comparison != 0) {
            return comparison;
        }
        return object.compareTo(other.object);
	}
	
	public Exp dir(Internable table) {
		return Pair.cons3(table.SUBJECT(), 
				table.PREDICATE(), 
				table.OBJECT(), 
				Pair.cons2(table.VARS(), table.CLASSES(), table.NIL()));
	}


}

File: src/org/genyris/dl/Triple.kt
package org.genyris.dl

import org.genyris.core.*
import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment

class Triple(val subject: Symbol, val predicate: Symbol, val `object`: Exp) : Atom(), Comparable<Any?> {
    @kotlin.Throws(GenyrisException::class)
    override fun acceptVisitor(guest: Visitor) {
        guest.visitTriple(this)
    }

    @kotlin.Throws(GenyrisException::class)
    override fun makeEnvironment(parent: Environment?): Environment {
        return TripleEnvironment(parent, this)
    }

    override fun toString(): String {
        return "(triple " + subject + " " + predicate + " " + `object` + ")"
    }

    override fun getBuiltinClassSymbol(table: Internable): Symbol? {
        return table.TRIPLE()
    }

    override fun hashCode(): Int {
        return subject.hashCode() + predicate.hashCode() + `object`.hashCode()
    }

    override fun equals(compare: Any?): Boolean {
        if (compare !is Triple) {
            return false
        }
        val t = compare
        return subject == t.subject && predicate == t.predicate && `object` == t.`object`
    }

    @kotlin.Throws(GenyrisException::class)
    override fun eval(env: Environment?): Exp {
        return this
    }

    override fun compareTo(O: Any?): Int {
        if (O !is Triple) {
            return -1
        }
        val other = O
        var comparison = subject.compareTo(other.subject)
        if (comparison != 0) {
            return comparison
        }
        comparison = predicate.compareTo(other.predicate)
        if (comparison != 0) {
            return comparison
        }
        return `object`.compareTo(other.`object`)
    }

    override fun dir(table: Internable): Exp {
        return Pair.Companion.cons3(
            table.SUBJECT(),
            table.PREDICATE(),
            table.OBJECT(),
            Pair.Companion.cons2(table.VARS(), table.CLASSES(), table.NIL())
        )
    }
}
